### PR TITLE
Homework 2

### PR DESCRIPTION
**Pull request by iRezograf**
Vladimir, HW2 is done. Approve it, please

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -13,27 +13,50 @@
...
-            <version>3.141.59</version>
-        </dependency -->
+            <version>${selenium.java.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.seleniumhq.webdriver</groupId>
+            <artifactId>webdriver-common</artifactId>
+            <version>${selenium.webdriver.common.version}</version>
+        </dependency>
```
Comment (shoggy) [inline pom.xml]: This is something veeery old. Please remove it

File: pom.xml [hunk 2]
```diff
@@ -13,27 +13,50 @@
...
             <artifactId>selenium-java</artifactId>
-            <version>3.141.59</version>
-        </dependency -->
+            <version>${selenium.java.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.seleniumhq.webdriver</groupId>
+            <artifactId>webdriver-common</artifactId>
+            <version>${selenium.webdriver.common.version}</version>
```
Comment (shoggy) [inline pom.xml]: please add `scope:test`

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+import ru.training.at.hw2.ex1.dataproviders.DpForLeftSectionTexts;
+import ru.training.at.hw2.ex1.dataproviders.DpForUnderFourIconsTexts;
+
+import javax.swing.text.html.HTMLDocument;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import static org.testng.AssertJUnit.assertEquals;
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: Please remove unused imports
If you use IntelliJ IDEA, the `Ctrl+Alt+O` shortcut can be used

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    public void tearDown() {
+        driver.quit();
+    }
+
+    @Test(groups = {"exercise_1"}, priority = 10)
+    public void openSiteByUrlTest() {
+        driver.navigate().to(testedSite);
+        driver.getTitle();
+    }
+
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java:56]: Please refactor test methods in the following way
1. create one method, annotate it with `@Test`
2. remove test annotation from all other methods
3. sequentially call old test methods from the new one
4. do not forget to check that `softAssert` does not contain any failures

something like
```java
@Test
public void someTestName() {
  openSiteByUrlTest();
  browserTitleTest();
  ...
  // check softAssert
}
```

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    SoftAssert softAssert;
+
+
+    String driverPath;
+    String testedSite;
+
+    @BeforeClass(groups = {"exercise_1"})
+    public void setUp() {
+        // Версия 90.0.4430.93
+        driverPath = "C:\\Users\\rezog\\driver\\chromedriver.exe";
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: Test is machine-dependent
Consider the use of bonigarcia webdrivermanager or put chromedriver in the project

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        element.sendKeys("Jdi1234");
+
+        element = driver.findElement(By.id("login-button"));
+        element.click();
+    }
+
+    @Test(groups = {"exercise_1"}, priority = 40)
+    public void usernameIsLogginedTest() {
+        element = driver.findElement(By.id("user-name"));
+        ExpectedConditions.textToBePresentInElement(element, "ROMAN IOVLEV!");
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: I think this statement is extra here, please remove it

This statement creates `ExpectedCondition` object that can be used in `org.openqa.selenium.support.ui.Wait` (explicit waits) but do not wait itself

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+        softAssert.assertEquals(element.getText(), "ROMAN IOVLEV");
+    }
+
+    @Test(groups = {"exercise_1"}, priority = 50)
+    public void headerHaveProperTextsTest() {
+
+        List<WebElement> elements = driver.findElements(By.tagName("li")).subList(0, 3);
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: Why it is a sublist?

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+        softAssert.assertEquals(elements.get(0).getText(), "HOME");
+        softAssert.assertEquals(elements.get(1).getText(), "CONTACT FORM");
+        softAssert.assertEquals(elements.get(2).getText(), "SERVICE");
+
+        element = driver.findElement(By.cssSelector(
+                "ul.uui-navigation.nav.navbar-nav.m-l8 >li:nth-child(4)"));
+        softAssert.assertEquals(element.getText(), "METALS & COLORS");
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: Why this item selector significantly differs from the other three?

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+        // Alternative method
+        element = driver.findElement(By.xpath(
+                "/html/body/header/div/nav/ul[1]/li[4]/a"));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: Please try to not use generated locators
locator that you used for "Metal and Colors" is almost right, correct it a bit

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        element = driver.findElement(By.cssSelector(
+                ".row.clerafix.benefits>.col-sm-3:nth-child(" + index + ")"));
+        softAssert.assertEquals(element.getText(),
+                expected, "Element's text isn't proper or not found");
+    }
+
+    @Test(groups = {"exercise_1"}, priority = 80)
+    public void iframeWithFrameButtonExistTest() {
+        String iframeWithFrameButton = "//iframe[@id='frame']";
+        WebElement iframe = driver.findElement(By.xpath(iframeWithFrameButton));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java:133]: Please use id `frame` or CSS `iframe#frame` for By definition and search of element

File: src/test/java/ru/training/at/hw2/ex2/ExerciseTwo.java [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        element = driver.findElement(By.cssSelector(
+                "ul.uui-navigation.nav.navbar-nav.m-l8>li.dropdown>ul li:nth-child(8)"));
+        element.click();
+    }
+
+    @Test(groups = {"exercise_2"}, priority = 155)
+    public void checkElementsTest() {
+        //Select checkboxes Water
+        element = driver.findElement(By.cssSelector(
+        "div.main-content > div > div:nth-child(2) >label:nth-child(1)"));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex2/ExerciseTwo.java]: Try to not use indices in selectors (sometimes it is acceptable, but in most cases, they can be avoided by use of the other selector e.g. by specifying class, tag ,input type)

**Review by shoggy — CHANGES_REQUESTED** [review]
Please apply comments to all applicable places (I tried to put one comment for the first appearance of the issue, but other appearances should be fixed in the same way)

File: pom.xml (hunk 1 above)
Comment (iRezograf) [reply pom.xml]: Done

File: pom.xml (hunk 2 above)
Comment (iRezograf) [reply pom.xml]: I have deleted it

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java (hunk 3 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: Done

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java (hunk 5 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: Done

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java (hunk 7 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: Fixed with:
element = driver.findElement(By.cssSelector(
                "ul.uui-navigation.nav.navbar-nav.m-l8 >li:nth-child(1)"));
        Assert.assertEquals(element.getText(), "HOME");
        Assert.assertTrue(element.isDisplayed());

        element = driver.findElement(By.cssSelector(
                "ul.uui-navigation.nav.navbar-nav.m-l8 >li:nth-child(2)"));
        Assert.assertEquals(element.getText(), "CONTACT FORM");
        Assert.assertTrue(element.isDisplayed());

        element = driver.findElement(By.cssSelector(
                "ul.uui-navigation.nav.navbar-nav.m-l8 >li:nth-child(3)"));
        Assert.assertEquals(element.getText(), "SERVICE");
        Assert.assertTrue(element.isDisplayed());

        element = driver.findElement(By.cssSelector(
                "ul.uui-navigation.nav.navbar-nav.m-l8>li:nth-child(4)"));
        Assert.assertEquals(element.getText(), "METALS & COLORS");
        Assert.assertTrue(element.isDisplayed());

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java (hunk 10 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java:133]: Fixed with:
element = driver.findElement(By.cssSelector(
                "ul.uui-navigation.nav.navbar-nav.m-l8 >li:nth-child(1)"));
        Assert.assertEquals(element.getText(), "HOME");
        Assert.assertTrue(element.isDisplayed());

        element = driver.findElement(By.cssSelector(
                "ul.uui-navigation.nav.navbar-nav.m-l8 >li:nth-child(2)"));
        Assert.assertEquals(element.getText(), "CONTACT FORM");
        Assert.assertTrue(element.isDisplayed());

        element = driver.findElement(By.cssSelector(
                "ul.uui-navigation.nav.navbar-nav.m-l8 >li:nth-child(3)"));
        Assert.assertEquals(element.getText(), "SERVICE");
        Assert.assertTrue(element.isDisplayed());

        element = driver.findElement(By.cssSelector(
                "ul.uui-navigation.nav.navbar-nav.m-l8>li:nth-child(4)"));
        Assert.assertEquals(element.getText(), "METALS & COLORS");
        Assert.assertTrue(element.isDisplayed());

File: src/test/java/ru/training/at/hw2/ex2/ExerciseTwo.java (hunk 11 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw2/ex2/ExerciseTwo.java]: Done

**Comment by iRezograf** [comment]
I fixed all remarks.

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java (hunk 6 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java]: Deleted

File: src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        Assert.assertEquals(element.getText(), dp.getUserNameAfterLogged());
+    }
+
+    public void headerHaveProperTextsTest() {
+        for (int i = 1; i <= 4; i++) {
+            element = driver.findElement(By.cssSelector(
+                    "ul.uui-navigation.nav.navbar-nav.m-l8 >li:nth-child(" + i + ")"));
+            Assert.assertEquals(element.getText(),
+                    dp.getTextOfHeaderMenuButtonsList().get(i - 1));
+        }
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseOne.java:89]: Would be better to use not hardcoded values for iteration
```suggestion
        for (int i = 0; i < getTextOfHeaderMenuButtonsList().size(); i++) {
            element = driver.findElement(By.cssSelector(
                    "ul.uui-navigation.nav.navbar-nav.m-l8 >li:nth-child(" + (i + 1) + ")"));
            Assert.assertEquals(element.getText(),
                    dp.getTextOfHeaderMenuButtonsList().get(i));
        }
```

\+ same for `fourTextsOnTheIndexPageUnderIconsAndTheyHaveProperTextTest` and `thereAreFiveItemsInTheLeftSectionTest()`

File: src/test/java/ru/training/at/hw2/ex2/ExerciseTwo.java [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+    public void openMenuServiceDifferentElementsTest() {
+        // Open drop down menu SERVICE
+        element = driver.findElement(By.cssSelector(
+                "ul.uui-navigation.nav.navbar-nav.m-l8>li.dropdown"));
+        element.click();
+
+        //8th p. of menu DIFFERENT ELEMENTS choose
+        element = driver.findElement(By.cssSelector(
+                "ul.uui-navigation.nav.navbar-nav.m-l8>li.dropdown>ul li:nth-child(8)"));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex2/ExerciseTwo.java:87]: In this case would be better to discard the use of indices, and replace it with search by text or by href 
```suggestion
                "ul.uui-navigation.nav.navbar-nav.m-l8>li.dropdown>ul li a[href='different-elements.html']"));
```

File: src/test/java/ru/training/at/hw2/ex2/ExerciseTwo.java [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+        element.click();
+    }
+
+    public void checkElementsTest() {
+        //Select checkboxes Water
+        //label.label-checkbox - it is checkbox
+        //label:nth-child(1) - it are elements of
+        element = driver.findElement(By.cssSelector(
+                "div.main-content > div > div:nth-child(2)>label:nth-child(1)"));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex2/ExerciseTwo.java]: At least block of inputs should be found without indices

**Review by shoggy — CHANGES_REQUESTED** [review]

**Review by shoggy — APPROVED** [review]

**Pull request merged**